File: apps/demo-be-jasb/src/main/java/com/organiclever/be/expense/repository/ExpenseRepository.java
package com.organiclever.be.expense.repository;

import com.organiclever.be.auth.model.User;
import com.organiclever.be.expense.model.Expense;
import java.util.Optional;
import java.util.UUID;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;

public interface ExpenseRepository extends JpaRepository<Expense, UUID> {
    Page<Expense> findAllByUser(User user, Pageable pageable);

    Optional<Expense> findByIdAndUser(UUID id, User user);
}

File: apps/demo-be-ktkt/src/main/kotlin/com/organiclever/demoktkt/infrastructure/repositories/ExpenseRepository.kt
package com.organiclever.demoktkt.infrastructure.repositories

import com.organiclever.demoktkt.domain.EntryType
import com.organiclever.demoktkt.domain.Expense
import com.organiclever.demoktkt.domain.Page
import java.math.BigDecimal
import java.time.LocalDate
import java.util.UUID

data class CreateExpenseRequest(
  val userId: UUID,
  val type: EntryType,
  val amount: BigDecimal,
  val currency: String,
  val category: String,
  val description: String,
  val date: LocalDate,
  val quantity: BigDecimal?,
  val unit: String?,
)

data class UpdateExpenseRequest(
  val type: EntryType,
  val amount: BigDecimal,
  val currency: String,
  val category: String,
  val description: String,
  val date: LocalDate,
  val quantity: BigDecimal?,
  val unit: String?,
)

data class CurrencySummary(val currency: String, val total: BigDecimal)

interface ExpenseRepository {
  suspend fun create(request: CreateExpenseRequest): Expense

  suspend fun findById(id: UUID): Expense?

  suspend fun findAllByUser(userId: UUID, page: Int, pageSize: Int): Page<Expense>

  suspend fun update(id: UUID, request: UpdateExpenseRequest): Expense?

  suspend fun delete(id: UUID): Boolean

  suspend fun summaryByUser(userId: UUID): List<CurrencySummary>

  suspend fun findByUserAndPeriod(
    userId: UUID,
    from: LocalDate,
    to: LocalDate,
    currency: String,
  ): List<Expense>
}
